contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,817 @@
+
+        return runningTotal;
+    }
+
+    /**
+     * Allows lenders to withdraw heldToken in the event that the position was force-recovered.
+     *
+     * @param  buckets  The bucket numbers to withdraw from
+     * @return          The number of heldTokens withdrawn
+     */
+    function withdrawHeldToken(
+        uint256[] buckets
+    )
+        external
+        returns(uint256)
+    {
+        // running total amount of tokens to withdraw
+        uint256 runningUserPrincipal = 0;
+        uint256 originalPrincipalTotal = principalTotal;
+
+        for (uint256 i = 0; i < buckets.length; i = i.add(1)) {
+            uint256 bucket = buckets[i];
+
+            // calculate the user's share
+            uint256 bucketWeight = weightForBkt[bucket];
+            uint256 userWeight = accountForWithdraw(bucket, msg.sender); // deletes the users share
+
+            // calculate the user's principal for the bucket
+            uint256 userPrincipal =
+                userWeight
+                .mul(principalForBkt[bucket])
+                .div(bucketWeight);
+
+            // update amounts
+            changePrincipal(bucket, userPrincipal, false);
+
+            runningUserPrincipal = runningUserPrincipal.add(userPrincipal);
+        }
+
+        uint256 tokenAmount = MathHelpers.getPartialAmount(
+            runningUserPrincipal,
+            originalPrincipalTotal,
+            TokenInteract.balanceOf(HELD_TOKEN, address(this))
+        );
+
+        TokenInteract.transfer(HELD_TOKEN, msg.sender, tokenAmount);
+
+        return tokenAmount;
+    }
+
+    // ============ Helper Functions ============
+
+    /**
+     * Updates the state variables at any time. Only does anything after the position has been
+     * closed or partially-closed since the last time this function was called.
+     *
+     * - Increases the available amount in the highest bucket with outstanding principal
+     * - Decreases the principal amount in that bucket
+     *
+     * @param  principalRemoved  Amount of principal closed since the last update
+     */
+    function accountForClose(
+        uint256 principalRemoved
+    )
+        internal
+    {
+        if (principalRemoved == 0) {
+            return;
+        }
+
+        uint256 newRepaidAmount = Margin(DYDX_MARGIN).getTotalOwedTokenRepaidToLender(POSITION_ID);
+        assert(newRepaidAmount.sub(cachedRepaidAmount) >= principalRemoved);
+
+        // find highest bucket with outstanding principal
+        uint256 bucket = getBucketNumber();
+        while (principalForBkt[bucket] == 0) {
+            bucket = bucket.sub(1);
+        }
+
+        // (available up / principal down) starting at the highest bucket
+        uint256 p_total = principalRemoved;
+        uint256 a_total = newRepaidAmount.sub(cachedRepaidAmount);
+        while (p_total > 0) {
+            uint256 p_i = Math.min256(p_total, principalForBkt[bucket]);
+            if (p_i == 0) {
+                continue;
```

Comment:
I'm confused, if `p_i` is 0 won't this infinite loop? did you mean to `break` instead?